### PR TITLE
Skip observation extraction in terminal states

### PR DESCRIPTION
**Pull request by dchetelat**
This pull request avoids computing the observation function in terminal states.
Following internal discussion, it also implements the "Justin trick" of temporarily removing time limits while extracting information, for making these steps atomic. (Not currently an issue but could be?)
Finally, this PR makes StrongBranchingScores return a vector of size `nb_vars` rather than `nb_cols`, which was an issue detected while investigating the above issue.

### CODE REVIEW

**Comment by dchetelat** [comment]
@AntoinePrv I'll let you review the code, maybe you want some stylistic changes?
Also, I'll run the code in parallel for the ABB project and I'll check if it solves the segfault, it should.

File: libecole/src/scip/model.cpp [hunk 1]
```diff
@@ -185,6 +216,30 @@ std::map<std::string, Param> Model::get_params() const {
...
+		"limits/bestsol",
+		"limits/maxsol",
+		"limits/maxorigsol",
+		"limits/restarts",
+		"limits/autorestartnodes",
+	};
+	auto saved = std::map<std::string, Param>{};
+	for (auto const& name : names) {
+		saved[name] = get_param<Param>(name);
+		set_param(name, get_param_default<Param>(name));
```
Comment (gasse) [inline libecole/src/scip/model.cpp:238]: Not sure those make a lot of sense:
```
"limits/nodes",
"limits/totalnodes",
"limits/stallnodes",
"limits/gap",
"limits/absgap",
"limits/solutions",
"limits/bestsol",
"limits/maxsol",
"limits/maxorigsol",
"limits/restarts",
"limits/autorestartnodes",
```
These parameters can stop / restart SCIP only when the state of the solver changes (node creation, solution creation etc.). We want to prevent SCIP to stop from within an observation function. Observation functions should not change the state of the solver, except maybe allocating some memory. There are other parameters, however, which we might want to manually set / restore, before / after observation functions that require to solve LPs are computed ?
```
# iteration limit for each single LP solve (-1: no limit)
# [type: longint, advanced: TRUE, range: [-1,9223372036854775807], default: -1]
lp/iterlim = -1

# LP algorithm for solving initial LP relaxations (automatic 's'implex, 'p'rimal simplex, 'd'ual simplex, 'b'arrier, barrier with 'c'rossover)
# [type: char, advanced: FALSE, range: {spdbc}, default: s]
lp/initalgorithm = s

# LP algorithm for resolving LP relaxations if a starting basis exists (automatic 's'implex, 'p'rimal simplex, 'd'ual simplex, 'b'arrier, barrier with 'c'rossover)
# [type: char, advanced: FALSE, range: {spdbc}, default: s]
lp/resolvealgorithm = s

# LP pricing strategy ('l'pi default, 'a'uto, 'f'ull pricing, 'p'artial, 's'teepest edge pricing, 'q'uickstart steepest edge pricing, 'd'evex pricing)
# [type: char, advanced: FALSE, range: {lafpsqd}, default: l]
lp/pricing = l

# should lp state be cleared at the end of probing mode when lp was initially unsolved, e.g., when called right after presolving?
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: TRUE]
lp/clearinitialprobinglp = TRUE

# should the LP be resolved to restore the state at start of diving (if FALSE we buffer the solution values)?
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: FALSE]
lp/resolverestore = FALSE

# should the buffers for storing LP solution values during diving be freed at end of diving?
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: FALSE]
lp/freesolvalbuffers = FALSE

# maximum age a dynamic column can reach before it is deleted from the LP (-1: don't delete columns due to aging)
# [type: int, advanced: TRUE, range: [-1,2147483647], default: 10]
lp/colagelimit = 10

# maximum age a dynamic row can reach before it is deleted from the LP (-1: don't delete rows due to aging)
# [type: int, advanced: TRUE, range: [-1,2147483647], default: 10]
lp/rowagelimit = 10

# should new non-basic columns be removed after LP solving?
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: FALSE]
lp/cleanupcols = FALSE

# should new basic rows be removed after LP solving?
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: TRUE]
lp/cleanuprows = TRUE

# should LP solver's return status be checked for stability?
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: TRUE]
lp/checkstability = TRUE

# maximum condition number of LP basis counted as stable (-1.0: no limit)
# [type: real, advanced: TRUE, range: [-1,1.79769313486232e+308], default: -1]
lp/conditionlimit = -1

# minimal Markowitz threshold to control sparsity/stability in LU factorization
# [type: real, advanced: TRUE, range: [0.0001,0.9999], default: 0.01]
lp/minmarkowitz = 0.01

# should LP solutions be checked for primal feasibility, resolving LP when numerical troubles occur?
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: TRUE]
lp/checkprimfeas = TRUE

# should LP solutions be checked for dual feasibility, resolving LP when numerical troubles occur?
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: TRUE]
lp/checkdualfeas = TRUE

# should infeasibility proofs from the LP be checked?
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: TRUE]
lp/checkfarkas = TRUE

# which FASTMIP setting of LP solver should be used? 0: off, 1: low
# [type: int, advanced: TRUE, range: [0,1], default: 1]
lp/fastmip = 1

# LP scaling (0: none, 1: normal, 2: aggressive)
# [type: int, advanced: TRUE, range: [0,2], default: 1]
lp/scaling = 1

# should presolving of LP solver be used?
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: TRUE]
lp/presolving = TRUE

# should the lexicographic dual algorithm be used?
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: FALSE]
lp/lexdualalgo = FALSE

# should the lexicographic dual algorithm be applied only at the root node
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: TRUE]
lp/lexdualrootonly = TRUE

# maximum number of rounds in the lexicographic dual algorithm (-1: unbounded)
# [type: int, advanced: TRUE, range: [-1,2147483647], default: 2]
lp/lexdualmaxrounds = 2

# choose fractional basic variables in lexicographic dual algorithm?
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: FALSE]
lp/lexdualbasic = FALSE

# turn on the lex dual algorithm only when stalling?
# [type: bool, advanced: TRUE, range: {TRUE,FALSE}, default: TRUE]
lp/lexdualstalling = TRUE

# disable the cutoff bound in the LP solver? (0: enabled, 1: disabled, 2: auto)
# [type: int, advanced: TRUE, range: [0,2], default: 2]
lp/disablecutoff = 2

# simplex algorithm shall use row representation of the basis if number of rows divided by number of columns exceeds this value (-1.0 to disable row representation)
# [type: real, advanced: TRUE, range: [-1,1.79769313486232e+308], default: 1.2]
lp/rowrepswitch = 1.2

# number of threads used for solving the LP (0: automatic)
# [type: int, advanced: TRUE, range: [0,64], default: 0]
lp/threads = 0

# factor of average LP iterations that is used as LP iteration limit for LP resolve (-1: unlimited)
# [type: real, advanced: TRUE, range: [-1,1.79769313486232e+308], default: -1]
lp/resolveiterfac = -1

# minimum number of iterations that are allowed for LP resolve
# [type: int, advanced: TRUE, range: [1,2147483647], default: 1000]
lp/resolveitermin = 1000

# LP solution polishing method (0: disabled, 1: only root, 2: always, 3: auto)
# [type: int, advanced: TRUE, range: [0,3], default: 3]
lp/solutionpolishing = 3

# LP refactorization interval (0: auto)
# [type: int, advanced: TRUE, range: [0,2147483647], default: 0]
lp/refactorinterval = 0

# should the Farkas duals always be collected when an LP is found to be infeasible?
# [type: bool, advanced: FALSE, range: {TRUE,FALSE}, default: FALSE]
lp/alwaysgetduals = FALSE
```

Comment (gasse) [inline libecole/src/scip/model.cpp:238]: Also, this code assumes that the default time limits are infinite. Is it always the case ?

File: libecole/src/observation/strongbranchingscores.cpp [hunk 2]
```diff
@@ -68,12 +68,12 @@ std::optional<xt::xtensor<double, 1>> StrongBranchingScores::extract(scip::Model
...
 	/* Store strong branching scores in tensor */
-	auto const num_lp_columns = static_cast<std::size_t>(SCIPgetNLPCols(scip));
-	auto strong_branching_scores = xt::xtensor<double, 1>({num_lp_columns}, std::nan(""));
+	auto const nb_vars = static_cast<std::size_t>(SCIPgetNVars(scip));
+	auto strong_branching_scores = xt::xtensor<double, 1>({nb_vars}, std::nan(""));
 
 	for (auto const [var, score] : views::zip(cands, cands_scores)) {
-		auto const lp_index = static_cast<std::size_t>(SCIPcolGetLPPos(SCIPvarGetCol(var)));
-		strong_branching_scores[lp_index] = static_cast<double>(score);
+		auto const var_index = static_cast<std::size_t>(SCIPvarGetProbindex(var));
```
Comment (gasse) [inline libecole/src/observation/strongbranchingscores.cpp:75]: This is a major change. For example, the branching tutorial will break. I suggest to make a separate PR for that.

File: python/src/ecole/environment.py [hunk 3]
```diff
@@ -170,9 +176,16 @@ def step(self, action, *dynamics_args, **dynamics_kwargs):
...
-            observation = self.observation_function.extract(self.model, done)
+
+            limit_values = self.model.pause_limits()
+            if not done:
+                observation = self.observation_function.extract(self.model, done)
+            else:
+                observation = None
             reward = self.reward_function.extract(self.model, done)
             information = self.information_function.extract(self.model, done)
+            self.model.set_params(limit_values)
```
Comment (gasse) [inline python/src/ecole/environment.py:187]: I would also do a separate PR to discuss this pause limits / restore limits thing. I can see a use case (the competition actually), where a reward function uses the "limits/solvingtime" parameter. I think it would be best to have each observation / reward / information function decide which parameters it wants to set / restore. At least, we should discuss it further.

**Review by gasse — CHANGES_REQUESTED** [review]
I think one contribution can be merged, and two others we should discuss more. I suggest to split the three contributions into three PRs.

File: libecole/src/observation/strongbranchingscores.cpp (hunk 2 above)
Comment (dchetelat) [reply libecole/src/observation/strongbranchingscores.cpp:75]: Why would the branching tutorial break?

File: libecole/src/scip/model.cpp (hunk 1 above)
Comment (dchetelat) [reply libecole/src/scip/model.cpp:238]: I checked and yes, except for two limits that, I hadn't realized, don't need to be included in the list, since they don't stop the solving.

```
# maximal number of solutions to store in the solution storage
# [type: int, advanced: FALSE, range: [1,2147483647], default: 100]
limits/maxsol = 100

# maximal number of solutions candidates to store in the solution storage of the original problem
# [type: int, advanced: FALSE, range: [0,2147483647], default: 10]
limits/maxorigsol = 10
```

File: libecole/src/observation/strongbranchingscores.cpp (hunk 2 above)
Comment (gasse) [reply libecole/src/observation/strongbranchingscores.cpp:75]: The tutorial assumes the scores array (SB or PseudoCost) is ordered according the LP index of the variables. Not their problem index. `scores[action_set]` does return the scores of the branching cancidates.
```
# We will solve problems (run episodes) until we have saved enough samples
max_samples_reached = False
while not max_samples_reached:
    episode_counter += 1
    
    observation, action_set, _, done, _ = env.reset(next(instances))
    while not done:
        (scores, scores_are_expert), node_observation = observation
        action = action_set[scores[action_set].argmax()]
```

File: libecole/src/scip/model.cpp (hunk 1 above)
Comment (dchetelat) [reply libecole/src/scip/model.cpp]: Moved to separate PR.

Comment (dchetelat) [reply libecole/src/scip/model.cpp]: Moved to separate PR.

File: python/src/ecole/environment.py (hunk 3 above)
Comment (dchetelat) [reply python/src/ecole/environment.py]: Moved to separate PR.

File: libecole/src/observation/strongbranchingscores.cpp (hunk 2 above)
Comment (dchetelat) [reply libecole/src/observation/strongbranchingscores.cpp]: I'm still not understanding you very well, but I checked and the action set returned by the Branching environment is indices w.r.t. column index
https://github.com/ds4dm/ecole/blob/7100c77a64389ab24e136e3cf0fec30850c88c4d/libecole/src/dynamics/branching.cpp#L17-L28
which is perhaps what you were trying to say? In any case that would be enough to break the tutorial, you are right.

This section has been moved to a separate PR.

File: libecole/include/ecole/environment/environment.hpp [hunk 4]
```diff
@@ -109,10 +116,18 @@ class Environment {
...
 			information_function().before_reset(model());
 			auto const [done, action_set] = dynamics().reset_dynamics(model(), std::forward<Args>(args)...);
-
 			can_transition = !done;
+
+			OptionalObservation observation;
+			if (can_transition) {
+				observation = observation_function().extract(model(), done);
+			} else if constexpr (is_optional_v<Observation>) {
+				observation = Observation{};
```
Comment (gasse) [inline libecole/include/ecole/environment/environment.hpp:125]: Can't you just return an OptionalObservation here ?

File: libecole/include/ecole/environment/environment.hpp [hunk 5]
```diff
@@ -109,10 +116,18 @@ class Environment {
...
+			} else {
+				observation = std::optional<Observation>{};
+			}
 			return {
-				observation_function().extract(model(), done),
+				observation,
```
Comment (gasse) [inline libecole/include/ecole/environment/environment.hpp:130]: I think std::move(observation) would make sense here ?

File: libecole/include/ecole/environment/environment.hpp [hunk 6]
```diff
@@ -109,10 +116,18 @@ class Environment {
 			reward_function().before_reset(model());
 			information_function().before_reset(model());
 			auto const [done, action_set] = dynamics().reset_dynamics(model(), std::forward<Args>(args)...);
-
 			can_transition = !done;
+
+			OptionalObservation observation;
+			if (can_transition) {
```
Comment (gasse) [inline libecole/include/ecole/environment/environment.hpp:122]: I would suggest `if (not done)` to keep the C++ code more in-sync with the python code.

File: libecole/src/observation/strongbranchingscores.cpp (hunk 2 above)
Comment (gasse) [reply libecole/src/observation/strongbranchingscores.cpp]: Yes that's what I meant, the tutorial code assumes column indices, so if we move to problem variable indices we'll have to fix the tutorial as well.

File: libecole/include/ecole/environment/environment.hpp (hunk 4 above)
Comment (dchetelat) [reply libecole/include/ecole/environment/environment.hpp:125]: Sorry, I'm not understanding you. What code would you like to see here?

Comment (gasse) [reply libecole/include/ecole/environment/environment.hpp:125]:
```suggestion
			if (can_transition) {
				observation = observation_function().extract(model(), done);
			} else {
				observation = OptionalObservation{};
			}
```

Comment (dchetelat) [reply libecole/include/ecole/environment/environment.hpp:125]: You mean this?
```c++
if (!done) {
    observation = observation_function().extract(model(), done);
} else {
    observation = OptionalObservation{};
}
```

Comment (dchetelat) [reply libecole/include/ecole/environment/environment.hpp:125]: In which case, I don't mind, if you think that's clearer. I think it's equivalent.

**Review by gasse — APPROVED** [review]

File: libecole/include/ecole/environment/environment.hpp (hunk 4 above)
Comment (dchetelat) [reply libecole/include/ecole/environment/environment.hpp]: Okay, changed that way.

File: libecole/include/ecole/environment/environment.hpp [hunk 7]
```diff
@@ -153,16 +166,23 @@ class Environment {
...
 		try {
 			auto const [done, action_set] = dynamics().step_dynamics(model(), action, std::forward<Args>(args)...);
 			can_transition = !done;
 
+			OptionalObservation observation;
+			if (!done) {
+				observation = observation_function().extract(model(), done);
+			} else {
+				observation = OptionalObservation{};
+			}
```
Comment (gasse) [inline libecole/include/ecole/environment/environment.hpp:183]: Same suggestion here:
```suggestion
			auto observation =  OptionalObservation{};
			if (!done) {
				observation = observation_function().extract(model(), done);
			}
```

File: libecole/include/ecole/environment/environment.hpp [hunk 8]
```diff
@@ -109,10 +116,16 @@ class Environment {
...
 			auto const [done, action_set] = dynamics().reset_dynamics(model(), std::forward<Args>(args)...);
-
 			can_transition = !done;
+
+			OptionalObservation observation;
+			if (!done) {
+				observation = observation_function().extract(model(), done);
+			} else {
+				observation = OptionalObservation{};
+			}
```
Comment (gasse) [inline libecole/include/ecole/environment/environment.hpp:126]: I think the `else` here is redundant, as `observation` is already instantiated to an empty observation when it is declared.. What about
```suggestion
			auto observation = OptionalObservation{};
			if (!done) {
				observation = observation_function().extract(model(), done);
			}
```

**Review by gasse — COMMENTED** [review]
Minor suggestions. Otherwise, good to go !

**Review by dchetelat — COMMENTED** [review]
It's probably ok here, since the observation constructors I think don't really do much, but this is actually less efficient... If `!done` you are calling the first constructor for nothing.

**Comment by gasse** [comment]
Thanks !
